File: packages/shared-types/src/map.ts
// Types for the digital floor map (nodes, edges, POIs, and status).
// Filled in on Day 2 when the data schemas are defined.
export type MapStatus = "draft" | "verified";

export type MapNode = {
  node_id: string;
  x: number;
  y: number;
  floor_number: number;
  kind: "corridor" | "room" | "elevator" | "stairs";
  location_id?: string | null;
  poi_id?: string | null;
  label?: string | null;
};

export type MapEdge = {
  edge_id: string;
  from_node: string;
  to_node: string;
  weight: number;
  kind: "walkway" | "door" | "elevator" | "stairs";
};

export type POI = {
  poi_id: string;
  node_id: string;
  location_id: string;
  label: string;
  x: number;
  y: number;
  floor_number: number;
};

export type MapFloor = {
  floor_id: string;
  floor_number: number;
  image_width: number;
  image_height: number;
  source_image: string;
};

export type DigitalMap = {
  map_id: string;
  hospital_id: string;
  building_id: string;
  status: MapStatus;
  floors: MapFloor[];
  nodes: MapNode[];
  edges: MapEdge[];
  pois: POI[];
  version: number;
  created_at: string;
  verified_at?: string | null;
};

export type MapListItem = {
  map_id: string;
  status: MapStatus;
  version: number;
  floor_count: number;
  node_count: number;
  edge_count: number;
  poi_count: number;
  created_at: string;
  verified_at?: string | null;
};

File: services/navigation-engine/app/models/map.py
"""Digital map models used by admin map builder and patient routing."""

from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel, Field


MapStatus = Literal["draft", "verified"]
NodeKind = Literal["corridor", "room", "elevator", "stairs"]
EdgeKind = Literal["walkway", "door", "elevator", "stairs"]


class MapFloor(BaseModel):
    floor_id: str
    floor_number: int
    image_width: int
    image_height: int
    source_image: str


class MapNode(BaseModel):
    node_id: str
    x: float
    y: float
    floor_number: int
    kind: NodeKind
    location_id: Optional[str] = None
    poi_id: Optional[str] = None
    label: Optional[str] = None


class MapEdge(BaseModel):
    edge_id: str
    from_node: str
    to_node: str
    weight: float
    kind: EdgeKind = "walkway"


class MapPOI(BaseModel):
    poi_id: str
    location_id: str
    node_id: str
    label: str
    x: float
    y: float
    floor_number: int


class DigitalMap(BaseModel):
    map_id: str
    hospital_id: str
    building_id: str
    status: MapStatus
    version: int = 1
    floors: list[MapFloor] = Field(default_factory=list)
    nodes: list[MapNode] = Field(default_factory=list)
    edges: list[MapEdge] = Field(default_factory=list)
    pois: list[MapPOI] = Field(default_factory=list)
    created_at: str
    verified_at: Optional[str] = None


class DigitizeMapRequest(BaseModel):
    map_id: Optional[str] = None
    floor_numbers: list[int] = Field(default_factory=lambda: [1, 2, 3])
    source_image_pattern: str = "map-floor{floor}.png"


class ConfirmMapRequest(BaseModel):
    map_id: str


class NodeAnchorUpdateRequest(BaseModel):
    x: float
    y: float
    status: MapStatus = "draft"


class RoutePreviewRequest(BaseModel):
    session_id: str | None = None
    start_location_id: str
    destination_location_id: str
    map_status: MapStatus = "draft"


class MapListItem(BaseModel):
    map_id: str
    status: MapStatus
    version: int
    floor_count: int
    node_count: int
    edge_count: int
    poi_count: int
    created_at: str
    verified_at: Optional[str] = None
